Add option to use NetCDF output instead of binary for point output

JessicaMeixner-NOAA opened the pull request:
# Pull Request Summary
Add a switch BIN2NC which will write a NetCDF file instead of a binary file for point output. 

## Description
When using BIN2NC, the point output will now be written in NetCDF from ww3_shel or ww3_multi.  The exes for output post-processing will be able to read these files.  Some notes: 
* ww3_ounp is still necessary.  While we can add at a future date direction and frequency so that programs such as ww3_bounc could use this without post-processing with ww3_ounf, there are many options and formats of variables that we likely do not want to be converting between in the model output.  Therefore, at this time, we are not replacing ww3_ounp.  There is room for future enhancement to reduce the need for ww3_ounp for some applications. 
* A switch was used instead of a namelist option because I suspect we will eventually remove binary, but in the meantime having a switch still allows those who wish to run without NetCDF that option.  I thought this was the simplest approach, but it's certainly not the only possibility and could be another future place for improvement. 

Co-authors are @edwardhartnett and @MatthewMasarik-NOAA 


Please also include the following information: 
* Mention any labels that should be added  _new feature_
* Are answer changes expected from this PR? While the format of the output will change and regtests that have added this feature will change, the other files do not change.  


### Issue(s) addressed

- fixes #682

### Commit Message
Add option to use NetCDF output instead of binary for point output 

### Check list  

<!-- After creating the PR you can check each of the items below that have been completed -->

- [x] Branch is up to date with the authoritative repository (NOAA-EMC) develop branch. 
- [x] Checked the [checklist for a developer submitting to develop](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-a-developer-submitting-to-develop). 
- [ ] If a version number update is required, checked the [updating version number](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-updating-version-number) checklist. 
- [x] If a new feature was added, a regression test for testing the new feature is added. 
 
### Testing

* How were these changes tested? With matrix.  Additionally, I hard-coded the switch option to always be on and then ran all of the regression tests using the NetCDF option (see branch: https://github.com/JessicaMeixner-NOAA/WW3/tree/feature/pointbinary2nc_hardcoded).  Output of this branch matrix comp is also shared below.  
* Are the changes covered by regression tests? (If not, why? Do new tests need to be added?) Yes, new tests were added. 
* Have the matrix regression tests been run (if yes, please note HPC and compiler)? Yes hera intel  (on this branch, plus the hard-coded branch). 
* Please indicate the expected changes in the regression test output, (Note the [list](https://github.com/NOAA-EMC/WW3/wiki/How-to-use-matrix.comp-to-compare-regtests-with-develop#4-look-at-results) of known non-identical tests.)

New tests: 
ww3 ww3_tp2.2/work_PR1_MPI_BIN2NC 
And updates to ww3_ufs1.1 unstructured tests as they will now use this feature.  Note the results do not chnage, just binary -> netcdf for the points, and the switch file in the netcdf output. 

* Please provide the summary output of matrix.comp (_matrix.Diff.txt_, _matrixCompFull.txt_ and _matrixCompSummary.txt_):


I ran two sets of tests, one from this branch compared with develop: 


```
**********************************************************************
********************* non-identical cases ****************************
**********************************************************************
mww3_test_03/./work_PR1_MPI_e                     (1 files differ)
mww3_test_03/./work_PR3_UQ_MPI_e_c                     (1 files differ)
mww3_test_03/./work_PR3_UNO_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UQ_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UNO_MPI_d2                     (16 files differ)
mww3_test_03/./work_PR1_MPI_d2                     (9 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2_c                     (11 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2_c                     (15 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2                     (15 files differ)
mww3_test_03/./work_PR2_UQ_MPI_d2                     (15 files differ)
mww3_test_03/./work_PR3_UQ_MPI_e                     (1 files differ)
mww3_test_03/./work_PR3_UNO_MPI_e_c                     (1 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2                     (16 files differ)
mww3_test_09/./work_MPI_ASCII                     (0 files differ)
ww3_tp2.10/./work_MPI_OMPH                     (7 files differ)
ww3_tp2.16/./work_MPI_OMPH                     (4 files differ)
ww3_tp2.2/./work_PR1_MPI_BIN2NC                     (directory not found)
ww3_tp2.6/./work_ST4_ASCII                     (0 files differ)
ww3_ufs1.1/./work_unstr_b                     (1 files differ)
ww3_ufs1.1/./work_unstr_a                     (1 files differ)
ww3_ufs1.1/./work_unstr_c                     (1 files differ)
ww3_ufs1.3/./work_a                     (3 files differ)
```



[matrixCompFull.txt](https://github.com/NOAA-EMC/WW3/files/15297702/matrixCompFull.txt)
[matrixCompSummary.txt](https://github.com/NOAA-EMC/WW3/files/15297703/matrixCompSummary.txt)
[matrixDiff.txt](https://github.com/NOAA-EMC/WW3/files/15297704/matrixDiff.txt)



And then I ran tests hard-coding the option to use BIN2NC by default (https://github.com/JessicaMeixner-NOAA/WW3/tree/feature/pointbinary2nc_hardcoded) to ensure all the various combinations were tested and ignoring "0 diffs" in the output the changes are as expected as well: 

```
**********************************************************************
********************* non-identical cases ****************************
**********************************************************************
mww3_test_03/./work_PR1_MPI_e                     (1 files differ)
mww3_test_03/./work_PR3_UQ_MPI_e_c                     (1 files differ)
mww3_test_03/./work_PR3_UNO_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UQ_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UNO_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UNO_MPI_d2                     (13 files differ)
mww3_test_03/./work_PR1_MPI_d2                     (21 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2_c                     (13 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2_c                     (13 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2                     (13 files differ)
mww3_test_03/./work_PR2_UQ_MPI_d2                     (14 files differ)
mww3_test_03/./work_PR3_UQ_MPI_e                     (1 files differ)
mww3_test_03/./work_PR3_UNO_MPI_e_c                     (1 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2                     (13 files differ)
ww3_tp2.10/./work_MPI_OMPH                     (5 files differ)
ww3_tp2.16/./work_MPI_OMPH                     (3 files differ)
ww3_tp2.2/./work_PR1_MPI_BIN2NC                     (directory not found)
ww3_ufs1.1/./work_unstr_b                     (1 files differ)
ww3_ufs1.1/./work_unstr_a                     (1 files differ)
ww3_ufs1.1/./work_unstr_c                     (1 files differ)
ww3_ufs1.3/./work_a                     (3 files differ)
```
* note the ww3_ufs1.1/./work_unstr_* cases are because the switch file changed in the NetCDF output. 


[matrixCompFull.txt](https://github.com/NOAA-EMC/WW3/files/15297692/matrixCompFull.txt)
[matrixCompSummary.txt](https://github.com/NOAA-EMC/WW3/files/15297693/matrixCompSummary.txt)
[matrixDiff.txt](https://github.com/NOAA-EMC/WW3/files/15297694/matrixDiff.txt)

edwardhartnett committed:
change to README

edwardhartnett committed:
adding w3iopomd

edwardhartnett committed:
cmake changes

edwardhartnett committed:
cmake changes

edwardhartnett committed:
cmake changes

edwardhartnett committed:
cmake changes

edwardhartnett committed:
cmake changes

edwardhartnett committed:
cmake changes

edwardhartnett committed:
clean up

edwardhartnett committed:
adding testing for binary restart file

edwardhartnett committed:
working on test

edwardhartnett committed:
merged develop

edwardhartnett committed:
merged

MatthewMasarik-NOAA committed:
ww3_outp.F90: hard-coded new subroutine (W3IOPON)

MatthewMasarik-NOAA committed:
gx_outp.F90 - W3IOPON

MatthewMasarik-NOAA committed:
w3wavemd.F90: W3IOPON

MatthewMasarik-NOAA committed:
wmiopomd.F90: W3IOPON

MatthewMasarik-NOAA committed:
ww3_ounp.F90: W3IOPON

MatthewMasarik-NOAA committed:
ww3_outp.F90: W3IOPON

MatthewMasarik-NOAA committed:
w3iopomd.F90: print statement

JessicaMeixner-NOAA committed:
small modifications to w3iopo for netcdf file name and removing print

JessicaMeixner-NOAA committed:
updates to documentation

JessicaMeixner-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/pointbinary2nc

JessicaMeixner-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/pointbinary2nc

JessicaMeixner-NOAA committed:
added notes for where time should be added and examined how
things varied from binary version to add warnings etc

JessicaMeixner-NOAA committed:
a first try at adding time dimension

edwardhartnett committed:
adding time dimension

edwardhartnett committed:
turning on tests

edwardhartnett committed:
Merge pull request #19 from edwardhartnett/ejh_0330_2

turning on tests

edwardhartnett committed:
more testing

edwardhartnett committed:
renamed file

edwardhartnett committed:
adding regtest document

edwardhartnett committed:
adding detail

edwardhartnett committed:
more info

edwardhartnett committed:
added tests to cmake file

edwardhartnett committed:
commenting out some of test

edwardhartnett committed:
commenting out some of test

edwardhartnett committed:
Merge pull request #20 from edwardhartnett/ejh_doc_0331

more testing

edwardhartnett committed:
added netCDF code

edwardhartnett committed:
uncommenting test code

edwardhartnett committed:
added reference

edwardhartnett committed:
more detail

edwardhartnett committed:
more detail

edwardhartnett committed:
Merge pull request #22 from edwardhartnett/ejh_0401

added netCDF code

edwardhartnett committed:
adding regtest CI

edwardhartnett committed:
working on regtest CI

edwardhartnett committed:
working on regtest CI

edwardhartnett committed:
working on regtest CI

edwardhartnett committed:
more regtest work

edwardhartnett committed:
more detail

edwardhartnett committed:
more details

edwardhartnett committed:
trying a regtest

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EH/ejh_test_ci' into feature/pointbinary2nc

 Conflicts:
	model/src/w3iopomd.F90

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/pointbinary2nc

JessicaMeixner-NOAA committed:
fix typo

edwardhartnett committed:
regtest progress

edwardhartnett committed:
regtest progress

edwardhartnett committed:
regtest progress

edwardhartnett committed:
more regtest

edwardhartnett committed:
more regtest

edwardhartnett committed:
more regtest

JessicaMeixner-NOAA committed:
fixes errors for successful build

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/pointbinary2nc

JessicaMeixner-NOAA committed:
save debug temporarily

JessicaMeixner-NOAA committed:
attempt to add start/count to netcdf putvar

edwardhartnett committed:
Merge pull request #23 from edwardhartnett/ejh_ci_regtest

adding regtest CI

edwardhartnett committed:
merged

edwardhartnett committed:
fixed merge problem

JessicaMeixner-NOAA committed:
try to add enddef to see if that fixes issue

JessicaMeixner-NOAA committed:
put enddef in correct location

edwardhartnett committed:
fixed merge problem

JessicaMeixner-NOAA committed:
inquire variable IDs if it's not the first time through the file

JessicaMeixner-NOAA committed:
fix for grdid

JessicaMeixner-NOAA committed:
can write file, adding write to check content

JessicaMeixner-NOAA committed:
fix output in write for dimensions

edwardhartnett committed:
fixing

edwardhartnett committed:
more detail

edwardhartnett committed:
merged

edwardhartnett committed:
reverted file

JessicaMeixner-NOAA committed:
add wdo to write and zero out II, IL, IW as in binary output

edwardhartnett committed:
fixed cmake

JessicaMeixner-NOAA committed:
add writing of cdo

JessicaMeixner-NOAA committed:
int -> float for appropriate variables

edwardhartnett committed:
pass tests

edwardhartnett committed:
Merge pull request #25 from edwardhartnett/JessicaMeixner-NOAA-feature/pointbinary2nc

Jessica meixner noaa feature/pointbinary2nc

JessicaMeixner-NOAA committed:
add time variable to output

JessicaMeixner-NOAA committed:
clean up of write debug statements and add a few comments

JessicaMeixner-NOAA committed:
remove the open and inquiries from the ipass=1 section of the read
-- we need to do this every time we enter this routine

JessicaMeixner-NOAA committed:
debugging tries for read routine

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/pointbinary2nc

edwardhartnett committed:
Merge pull request #21 from edwardhartnett/ejh_regtest_doc

adding regtest document

edwardhartnett committed:
porting changes

edwardhartnett committed:
merged Jessica's changes

JessicaMeixner-NOAA committed:
update documentation PTLOC is a real not int

edwardhartnett committed:
test writes netCDF file

edwardhartnett committed:
Merge pull request #27 from edwardhartnett/ejh_0404

porting changes

edwardhartnett committed:
minor change

JessicaMeixner-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/pointbinary2nc

JessicaMeixner-NOAA committed:
Merge branch 'feature/pointbinary2nc' of github.com:JessicaMeixner-NOAA/WW3 into feature/pointbinary2nc

JessicaMeixner-NOAA committed:
saving debug output write

edwardhartnett committed:
added expected output for regtest

edwardhartnett committed:
updated workflow

edwardhartnett committed:
updated workflow

edwardhartnett committed:
updated workflow

edwardhartnett committed:
Merge pull request #28 from edwardhartnett/ejh_0405

add expected output file for regtest ww3_tp2.5

edwardhartnett committed:
working on read code

edwardhartnett committed:
fixed test_io call of read

edwardhartnett committed:
Merge pull request #30 from edwardhartnett/ejh_0410

Ejh 0410

edwardhartnett committed:
merged develop

edwardhartnett committed:
testing

edwardhartnett committed:
Merge pull request #34 from edwardhartnett/ejh_r3

testing

edwardhartnett committed:
trying to cache regtest data

edwardhartnett committed:
fixing cache

edwardhartnett committed:
Merge pull request #35 from edwardhartnett/ejh_r4

trying to cache regtest data

edwardhartnett committed:
porting work

edwardhartnett committed:
fixed code

edwardhartnett committed:
more unit tests

edwardhartnett committed:
better error handling

edwardhartnett committed:
better error handling

edwardhartnett committed:
better error handling

edwardhartnett committed:
clean up regtest CI, revert change

edwardhartnett committed:
fixed error handling

edwardhartnett committed:
fixed comment

edwardhartnett committed:
Merge pull request #36 from edwardhartnett/ejh_r7

Ejh r7

edwardhartnett committed:
cause run_cmake_test to return error if there is an error

edwardhartnett committed:
cause run_cmake_test to return error if there is an error

edwardhartnett committed:
adding test_io3

edwardhartnett committed:
adding time dimension to read

edwardhartnett committed:
removing time dimension to read

edwardhartnett committed:
check for associated memory when reading

edwardhartnett committed:
initializing point data arrays on read

edwardhartnett committed:
more changes

JessicaMeixner-NOAA committed:
successfully reading in variables now!

JessicaMeixner-NOAA committed:
update read for time

edwardhartnett committed:
fixed inadvertant change

edwardhartnett committed:
Merge pull request #37 from edwardhartnett/ejh_0413

Ejh 0413

JessicaMeixner-NOAA committed:
add a flag to stop at end of file for netcdf point read

JessicaMeixner-NOAA committed:
small fix to make tp1.1 match develop

JessicaMeixner-NOAA committed:
clean up comments and print statements

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EH/ejh_develop' into feature/pointbinary2nc

 Conflicts:
	model/src/w3iopomd.F90

JessicaMeixner-NOAA committed:
fix bug from merge

JessicaMeixner-NOAA committed:
updates for FLX5 and SETUP

JessicaMeixner-NOAA committed:
fix build errors and a typo

JessicaMeixner-NOAA committed:
clean up if statements

JessicaMeixner-NOAA committed:
clean up test files

JessicaMeixner-NOAA committed:
remove duplication of wiki page

JessicaMeixner-NOAA committed:
update error messages to use write statements and add line numbers

JessicaMeixner-NOAA committed:
update error messages to show line number for easier debugging
following the updates from ww3_prnc

JessicaMeixner-NOAA committed:
fix for a few multi tests to run

JessicaMeixner-NOAA committed:
some clean up

JessicaMeixner-NOAA committed:
add a check on version numbers

JessicaMeixner-NOAA committed:
minor update

JessicaMeixner-NOAA committed:
test update to ci

JessicaMeixner-NOAA committed:
first step of adding BIN2NC switch

JessicaMeixner-NOAA committed:
add cmake things for bin2nc switch

JessicaMeixner-NOAA committed:
more updates for bin2nc

JessicaMeixner-NOAA committed:
update to have regtests pass

JessicaMeixner-NOAA committed:
add regtests

JessicaMeixner-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/pointbinary2nc

JessicaMeixner-NOAA committed:
get ci to work

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/pointbinary2nc

MatthewMasarik-NOAA commented:
@JessicaMeixner-NOAA it's great to see this worked PR'd!  I'll start reviewing.

JessicaMeixner-NOAA commented:
@mickaelaccensi Great question about timings on binary versus NetCDF.  Just using the regression tests you can see the differences in "Total Run Time" here: 
[TotalRunTimeBIN2NC.csv](https://github.com/NOAA-EMC/WW3/files/15310724/TotalRunTimeBIN2NC.csv)


This combines runtimes of all programs, so more detailed looks might be required.  This has four columns, the regtest name, the develop ranch, this PR, the hard-coded branch with writing netcdf for all points output.

JessicaMeixner-NOAA commented:
@mickaelaccensi I updated my last comment.  I don't see systematic differences in timing with using netcdf versus binary, but I'm also not sure these are the best tests to confirm this.  Let me know if you'd like to see something specific timing wise and I'll share if I get any additional results.

JessicaMeixner-NOAA commented:
@ukmo-ccbunney @mickaelaccensi just wanted to see if you'd had a chance to look at this at all and if you had any questions.

ukmo-ccbunney reviewed (commented):
I have tried running the new ww3_tp2.2 regression test and get the following error:
```
  WAVEWATCH III calculating for 2022/06/06 04:30:00 UTC at 11:25:07
  WAVEWATCH III calculating for 2022/06/06 05:00:00 UTC at 11:25:07
  WAVEWATCH III calculating for 2022/06/06 05:30:00 UTC at 11:25:07
  WAVEWATCH III calculating for 2022/06/06 06:00:00 UTC at 11:25:08
  WAVEWATCH III calculating for 2022/06/06 06:30:00 UTC at 11:25:08
application called MPI_Abort(MPI_COMM_WORLD, 21) - process 1
  *** WAVEWATCH III ERROR IN W3IOPO :
  LINE NUMBER         1556
  NETCDF ERROR MESSAGE: 
 Error: No such file or directory
  *** WAVEWATCH III ERROR IN W3IOPO :
  LINE NUMBER         1773
  NETCDF ERROR MESSAGE: 
 Error: No such file or directory
  *** WAVEWATCH III ERROR IN W3IOPO :
 Nonzero return at end of W3IOPON
 Error: No such file or directory
```

I am investigating.

ukmo-ccbunney commented:
The `filename` for netCDF output appears to contain junk or is getting corrupted:
```
 CB: Filename created: ./out_pnt.ww3.nc�`�
 CB Creating file: ./out_pnt.ww3.nc�`�
  WAVEWATCH III calculating for 2022/06/06 00:36:00 UTC at 12:40:14
  WAVEWATCH III calculating for 2022/06/06 01:12:00 UTC at 12:40:15
  WAVEWATCH III calculating for 2022/06/06 01:48:00 UTC at 12:40:15
  WAVEWATCH III calculating for 2022/06/06 02:24:00 UTC at 12:40:16
  WAVEWATCH III calculating for 2022/06/06 03:00:00 UTC at 12:40:16
```

when trying to repoen the file it gets a different filename:
```
  WAVEWATCH III calculating for 2022/06/06 06:30:00 UTC at 12:40:19
application called MPI_Abort(MPI_COMM_WORLD, 21) - process 1
 CB: Filename created: ./out_pnt.ww3.nc p�a/users/frey/WW3Git/JessicaWW3/regtests/ww3_tp2.2/work_PR1_MPI_BIN2NC/./ut_pnt.ww3.
 CB reopening file: ./out_pnt.ww3.nc p�a/users/frey/WW3Git/JessicaWW3/regtests/ww3_tp2.2/work_PR1_MPI_BIN2NC/./ut_pnt.ww3.
  *** WAVEWATCH III ERROR IN W3IOPO :
  LINE NUMBER         1558
  NETCDF ERROR MESSAGE: 
 Error: No such file or directory
```

I am not sure why this is happening at the moment.

ukmo-ccbunney commented on model/src/w3iopomd.F90:
```diff
@@ -1024,6 +1109,683 @@ SUBROUTINE W3IOPE ( A )
+    IF ( OFILES(2) .EQ. 1 ) THEN 
+      ! Create TIMETAG for file name using YYYYMMDD.HHMMS prefix
+      WRITE(TIMETAG,"(i8.8,'.'i6.6)")TIME(1),TIME(2)
+      filename = transfer(FNMPRE(:LEN_TRIM(FNMPRE))//TIMETAG//'out_pnt.'//FILEXT(:LEN_TRIM(FILEXT))//'.nc', filename)
```
These `TRANSFER` functions are causing filename corruption when I run on GNU (see my previous comments). If I remove the `TRANSFER` function and just build the filename with string concatenation it all works fine.

Is there a reason for the use of `TRANSFER` here? I can't see why it is needed. Could it be removed?

JessicaMeixner-NOAA replied:
@ukmo-ccbunney thank you for testing and finding this error! I'll change the way the filename is defined and have an update - hopefully by the end of the day for me, but likely tomorrow morning for you!

JessicaMeixner-NOAA replied:
@ukmo-ccbunney made an update that ran for me.  Hopefully this will work for you as well.  I just ran the one test, will start running more now.

JessicaMeixner-NOAA replied:
Okay my tests have all looked okay.  I did find one minor update for when we use the TIMETAG.  Let me know if things do not work for you with this latest change and I'll find a different solution @ukmo-ccbunney

ukmo-ccbunney replied:
@JessicaMeixner-NOAA - all working for me now in ww3_tp2.2.

MatthewMasarik-NOAA replied:
@JessicaMeixner-NOAA is this ready for me to test the update?

JessicaMeixner-NOAA replied:
@MatthewMasarik-NOAA Yes, this is ready for you to re-test!

MatthewMasarik-NOAA replied:
@JessicaMeixner-NOAA awesome, I'll get them started!

ukmo-ccbunney reviewed (changes requested):
TRANSFER function is causing problems in the construction of `filename`. It is resulting in non-reproducible garbage at the end of the filename.

JessicaMeixner-NOAA committed:
remove transfer function

mickaelaccensi commented:
I'm running the matrix. I'll keep you informed

JessicaMeixner-NOAA committed:
fix format when using timetag

JessicaMeixner-NOAA commented:
@ukmo-ccbunney @mickaelaccensi - Please let me know if you are running into any issues with this that I might be able to help with debugging.

mickaelaccensi commented:
matrix is done but I have both an error in develop and bin2nc branches for ww3_tp2.10 for output point processing due to the fact that the only requested output point (B45005) is considered on land. It is a SMC grid test so it is not linked to this PR. @ukmo-ccbunney and @ukmo-jianguo-li can probably double check this problem.

I've checked at ww3_tp2.2 for the format of the out_pnt.ww3.nc, it would be great to have at least the time variable defined in the correct calendar format and standard instead of a 2d array of integers. I think using the functions in w3timemd.ftn it should not bet too complicated..

Anyway, I'm now running the matrix.comp. Once done I'll keep you informed.

mickaelaccensi commented:
```
mww3_test_03/./work_PR2_UQ_MPI_d2                     (15 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2                     (12 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2_c                     (15 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2_c                     (14 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2                     (15 files differ)
mww3_test_03/./work_PR1_MPI_d2                     (14 files differ)
mww3_test_03/./work_PR3_UQ_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UNO_MPI_d2                     (15 files differ)
ww3_tp2.10/./work_MPI                     (1 files differ)
ww3_tp2.17/./work_mc1                     (6 files differ)
ww3_tp2.17/./work_mc                     (5 files differ)
ww3_tp2.17/./work_c                     (8 files differ)
ww3_tp2.17/./work_b                     (4 files differ)
ww3_tp2.17/./work_mb                     (3 files differ)
ww3_tp2.2/./work_PR1_MPI_BIN2NC                     (directory not found)
ww3_tp2.21/./work_b                     (6 files differ)
ww3_tp2.21/./work_mb                     (7 files differ)
ww3_tp2.6/./work_pdlib                     (14 files differ)
```

differences sound ok. tp2.17, ww2.21 and ww3_tp2.6 are due to pdlib not b4b

Except the format of time variable in out_pnt.ww3.nc, I'm ok with this PR

ukmo-jianguo-li commented:
This recent SMC grid update has activated an old conversion factor VSC to reverse the depth values stored in the SMC grid cell array before they are assigned to the WW3 ZB variable:

!/SMC !Li  Minimum DMIN depth is used as well for SMC.
!/SMC !!       ZB(ISEA)= - MAX( DMIN, FLOAT( IJKCel(5, ISEA) ) )
!/SMC !Li  Allow land cell to be defined by ZLIM value and only reset
!/SMC !Li  MAPST* land values for sea points.   JGLi03Nov2023
!/SMC          ZB(ISEA)= DVSMC*FLOAT( IJKCel(5, ISEA) )

The DVSMC is a copy of the VSC conversion factor, which is specified in the w3_grid.inp file on the old depth input file line:
-0.1  10.0  30   -1. 1 1 '(....)' 'NAME' 'S36125Depth.dat'
Where the fourth argument -1.(0) Is for the VSC.  This factor is not defined in the namelist input file and it results in all wet points are converted into land points.  The namelist input file in tp2.10 test needs to be updated for this factor.  If the old ww3_grid.inp file is used, it should be fine.


From: Mickael Accensi ***@***.***>
Sent: Friday, May 31, 2024 8:17 AM
To: NOAA-EMC/WW3 ***@***.***>
Cc: JianGuo Li ***@***.***>; Mention ***@***.***>
Subject: Re: [NOAA-EMC/WW3] Add option to use NetCDF output instead of binary for point output (PR #1230)


matrix is done but I have both an error in develop and bin2nc branches for ww3_tp2.10 for output point processing due to the fact that the only requested output point (B45005) is considered on land. It is a SMC grid test so it is not linked to this PR. @ukmo-ccbunney<https://github.com/ukmo-ccbunney> and @ukmo-jianguo-li<https://github.com/ukmo-jianguo-li> can probably double check this problem.

I've checked at ww3_tp2.2 for the format of the out_pnt.ww3.nc, it would be great to have at least the time variable defined in the correct calendar format and standard instead of a 2d array of integers. I think using the functions in w3timemd.ftn it should not bet too complicated..

Anyway, I'm now running the matrix.comp. Once done I'll keep you informed.

-
Reply to this email directly, view it on GitHub<https://github.com/NOAA-EMC/WW3/pull/1230#issuecomment-2141371942>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ANGV3EHWHKKWZLMEAUVVJOLZFAPXFAVCNFSM6AAAAABHUSEINOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCNBRGM3TCOJUGI>.
You are receiving this because you were mentioned.Message ID: ***@***.******@***.***>>

mickaelaccensi commented:
thanks @ukmo-jianguo-li  for the exclanation. I'll let you create an issue for this so.

JessicaMeixner-NOAA commented:
> Except the format of time variable in out_pnt.ww3.nc, I'm ok with this PR

@mickaelaccensi The time format is what WW3 is expecting to read in and then use in the next routines.  I'm thinking we just add an additional time variable  as you requested instead of changing this one, thoughts?

ukmo-ccbunney commented:
> thanks @ukmo-jianguo-li for the exclanation. I'll let you create an issue for this so.

Thanks Mickael - I'll raise an Issue for this.

mickaelaccensi commented:
> > Except the format of time variable in out_pnt.ww3.nc, I'm ok with this PR
> 
> @mickaelaccensi The time format is what WW3 is expecting to read in and then use in the next routines. I'm thinking we just add an additional time variable as you requested instead of changing this one, thoughts?

It seems to be a solution. To do it, you would need to rename the current time variable as another name (like time_ww3) and create a new time variable with the calendar format. It would be great for future use the out_grd.ww3.nc

JessicaMeixner-NOAA committed:
add a WW3time (which is whats read) and the more expected time var

JessicaMeixner-NOAA commented:
@mickaelaccensi and @ukmo-ccbunney I have added the extra time variable as @mickaelaccensi requested.    Let me know what you think.  @ukmo-ccbunney I think I have addressed your concerns of things not working, but I still see changes requested, so if I missed something or things are still not working for you please let me know!

mickaelaccensi commented:
@JessicaMeixner-NOAA, thanks first to have added the TIME variable. There is just an issue in the TIME values. See screenshot

![image](https://github.com/NOAA-EMC/WW3/assets/49198861/ed705146-4e66-4403-b88b-3c9ceabff654)

JessicaMeixner-NOAA commented:
Okay i just went to debug this and check this out and i somehow deleted a few lines of code.  Should have a fix here in an hour or so.  Thanks for catching this @mickaelaccensi !

JessicaMeixner-NOAA committed:
properly calculate time

JessicaMeixner-NOAA commented:
@mickaelaccensi  time should be fixed now!  Please let me know if you see issues still.

mickaelaccensi commented:
I confirm the bugfix !

JessicaMeixner-NOAA commented:
@ukmo-ccbunney - are things working for you on your end?

ukmo-ccbunney reviewed (approved):
Regression test `ww3_tp2.2/PR1_MPI_BIN2NC` ran fine and output looks good.
Running full regtest matrix now for completeness.

ukmo-ccbunney commented:
All regtests ran ok with just the usual differences in mww3_test_03. 👍

JessicaMeixner-NOAA commented:
Thanks @ukmo-ccbunney! 

@MatthewMasarik-NOAA - I think this is ready for your final review now!

MatthewMasarik-NOAA commented:
> Thanks @ukmo-ccbunney!
> 
> @MatthewMasarik-NOAA - I think this is ready for your final review now!

Great, thanks for confirming, I have the review in progress now.

MatthewMasarik-NOAA reviewed (approved):
**Code review**
Pass

**Testing**
Pass

* Compared with `develop` branch:
```
**********************************************************************
********************* non-identical cases ****************************
**********************************************************************
mww3_test_03/./work_PR1_MPI_e                     (1 files differ)
mww3_test_03/./work_PR3_UQ_MPI_e_c                     (1 files differ)
mww3_test_03/./work_PR3_UNO_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UQ_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UNO_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UNO_MPI_d2                     (17 files differ)
mww3_test_03/./work_PR1_MPI_d2                     (14 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2_c                     (13 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2_c                     (16 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2                     (15 files differ)
mww3_test_03/./work_PR2_UQ_MPI_d2                     (15 files differ)
mww3_test_03/./work_PR3_UNO_MPI_e_c                     (1 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2                     (15 files differ)
mww3_test_09/./work_MPI_ASCII                     (0 files differ)
ww3_tp2.10/./work_MPI_OMPH                     (7 files differ)
ww3_tp2.16/./work_MPI_OMPH                     (4 files differ)
ww3_tp2.6/./work_ST4_ASCII                     (0 files differ)
ww3_ufs1.1/./work_unstr_b                     (1 files differ)
ww3_ufs1.1/./work_unstr_a                     (1 files differ)
ww3_ufs1.1/./work_unstr_c                     (1 files differ)
ww3_ufs1.3/./work_a                     (1 files differ)
 
**********************************************************************
************************ identical cases *****************************
**********************************************************************
```
[matrixCompFull.txt](https://github.com/user-attachments/files/15841404/matrixCompFull.txt)
[matrixDiff.txt](https://github.com/user-attachments/files/15841407/matrixDiff.txt)
[matrixCompSummary.txt](https://github.com/user-attachments/files/15841408/matrixCompSummary.txt)


* Compared with `feature/pointbinary2nc_hardcoded` branch:

[hardcoded.matrixCompFull.txt](https://github.com/user-attachments/files/15841412/hardcoded.matrixCompFull.txt)
[hardcoded.matrixDiff.txt](https://github.com/user-attachments/files/15841413/hardcoded.matrixDiff.txt)
[hardcoded.matrixCompSummary.txt](https://github.com/user-attachments/files/15841414/hardcoded.matrixCompSummary.txt)


Approved.

MatthewMasarik-NOAA commented:
A big thank you to @JessicaMeixner-NOAA, who took over the reins and saw this work to completion, and @edwardhartnett who initiated the work and laid the foundation.  This is a great contribution in converting WW3 binary outputs to netcdf!  Thanks also goes to @mickaelaccensi and @ukmo-ccbunney for testing, and giving feedback that improved the final PR.
